lib/dialects/postgres/query-interface.js

```diff
@@ -2,154 +2,252 @@
+    const queryOptions = {
+      ...options,
+      type: QueryTypes.FOREIGNKEYS
+    };
 
-  // If ENUM processed, then refresh OIDs
-  if (promises.length) {
-    await qi.sequelize.dialect.connectionManager._refreshDynamicOIDs();
+    // postgres needs some special treatment as those field names returned are all lowercase
+    // in order to keep same result with other dialects.
+    const query = this.queryGenerator.getForeignKeyReferencesQuery(tableName, this.sequelize.config.database);
+    const result = await this.sequelize.query(query, queryOptions);
+    return result.map(Utils.camelizeObjectKeys);
+  }
+
+  /**
+   * Drop specified enum from database (Postgres only)
+   *
+   * @param {string} [enumName]  Enum name to drop
+   * @param {object} options Query options
+   *
+   * @returns {Promise}
+   */
+  async dropEnum(enumName, options) {
+    options = options || {};
+
+    return this.sequelize.query(
+      this.queryGenerator.pgEnumDrop(null, null, this.queryGenerator.pgEscapeAndQuote(enumName)),
+      { ...options, raw: true }
+    );
+  }
+
+  /**
+   * Drop all enums from database (Postgres only)
+   *
+   * @param {object} options Query options
+   *
+   * @returns {Promise}
+   */
+  async dropAllEnums(options) {
+    options = options || {};
+
+    const enums = await this.pgListEnums(null, options);
+
+    return await Promise.all(enums.map(result => this.sequelize.query(
+      this.queryGenerator.pgEnumDrop(null, null, this.queryGenerator.pgEscapeAndQuote(result.enum_name)),
+      { ...options, raw: true }
+    )));
+  }
+
+  /**
+   * List all enums (Postgres only)
+   *
+   * @param {string} [tableName]  Table whose enum to list
+   * @param {object} [options]    Query options
+   *
+   * @returns {Promise}
+   */
+  async pgListEnums(tableName, options) {
+    options = options || {};
+    const sql = this.queryGenerator.pgListEnums(tableName);
+    return this.sequelize.query(sql, { ...options, plain: false, raw: true, type: QueryTypes.SELECT });
   }
-  return result;
-}
 
+  /**
+   * Since postgres has a special case for enums, we should drop the related
+   * enum type within the table and attribute
+   *
+   * @override
+   */
+  async dropTable(tableName, options) {
+    await super.dropTable(tableName, options);
+    const promises = [];
+    const instanceTable = this.sequelize.modelManager.getModel(tableName, { attribute: 'tableName' });
+
+    if (!instanceTable) {
+      // Do nothing when model is not available
+      return;
+    }
+
+    const getTableName = (!options || !options.schema || options.schema === 'public' ? '' : `${options.schema}_`) + tableName;
+
+    const keys = Object.keys(instanceTable.rawAttributes);
+    const keyLen = keys.length;
+
+    for (let i = 0; i < keyLen; i++) {
+      if (instanceTable.rawAttributes[keys[i]].type instanceof DataTypes.ENUM) {
+        const sql = this.queryGenerator.pgEnumDrop(getTableName, keys[i]);
+        options.supportsSearchPath = false;
```

Comment:
@SimonSchick @sushantdhiman 

Looks like `options` are optional here, so if you invoke Model.drop without options, there will be an error here.

```
TypeError: Cannot set property 'supportsSearchPath' of undefined
    at PostgresQueryInterface.dropTable (node_modules/sequelize/lib/dialects/postgres/query-interface.js:237:36)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```